c++/src/kj/compat/http.c++

```diff
@@ -3677,6 +4027,109 @@ public:
               });
             }
 
+            kj::Maybe<CompressionParameters> compressionParameters;
+            KJ_IF_MAYBE(agreedParameters, responseHeaders.get(
+                HttpHeaderId::SEC_WEBSOCKET_EXTENSIONS)) {
+              kj::StringPtr failed = "Server failed WebSocket handshake: "_kj;
+              // If the response included websocket extensions, let's make sure we only received a
+              // single (valid) `permessage-deflate` extension.
+              if (offeredExtensions == nullptr) {
+                // We've received extensions when we did not send any in the first place.
+                auto msg = kj::str(
+                    failed, "added Sec-WebSocket-Extensions when client did not offer any.");
+                return settings.errorHandler.orDefault(*this).handleWebSocketProtocolError({
+                  502, "Bad Gateway", msg, nullptr
+                });
+              }
+              auto offers = _::splitPartsWithCondition(*agreedParameters, ',');
```

Comment:
There's a whole bunch of string twiddling here that feels redundant.

Couldn't you reuse `retainClientCompressorConfig()` (or a very similar function) to parse the response from the server, and then compare the parsed structures, rather than carefully pick through the text again here?

Reply:
https://github.com/capnproto/capnproto/pull/1480#discussion_r910503319

Really boils down to: client expects one config that the server agreed to (agreed parameters), and if the client declines this agreement we have to completely fail the connection. OTOH if the server declined _all_ the extensions that the client offered, it would still return a websocket, though the response would not include the `Sec-WebSocket-Extensions` header. 

The relevant commit to this comment is the middle of the three in the comment I linked.